AndroidSDK/src/com/leanplum/LeanplumGcmProvider.java

```diff
@@ -117,52 +108,62 @@ public String getRegistrationId() {
-
-    boolean hasGcmReceiver = LeanplumManifestHelper.checkComponent(
-        LeanplumManifestHelper.getReceivers(), GCM_RECEIVER, true, SEND_PERMISSION,
-        Arrays.asList(RECEIVE_ACTION, REGISTRATION_ACTION), context.getPackageName());
-    boolean hasPushReceiver = LeanplumManifestHelper.checkComponent(
-        LeanplumManifestHelper.getReceivers(), PUSH_RECEIVER, false, null,
-        Collections.singletonList(PUSH_LISTENER_SERVICE), null);
-
-    boolean hasReceivers = hasGcmReceiver && hasPushReceiver;
-
-    boolean hasPushListenerService = LeanplumManifestHelper.checkComponent(
-        LeanplumManifestHelper.getServices(), PUSH_LISTENER_SERVICE, false, null,
-        Collections.singletonList(RECEIVE_ACTION), null);
-    boolean hasPushInstanceIDService = LeanplumManifestHelper.checkComponent(
-        LeanplumManifestHelper.getServices(), PUSH_INSTANCE_ID_SERVICE, false, null,
-        Collections.singletonList(INSTANCE_ID_ACTION), null);
-    boolean hasPushRegistrationService = LeanplumManifestHelper.checkComponent(
-        LeanplumManifestHelper.getServices(), PUSH_REGISTRATION_SERVICE, false, null, null, null);
-
-    boolean hasServices = hasPushListenerService && hasPushInstanceIDService
-        && hasPushRegistrationService;
-
-    return hasPermissions && hasReceivers && hasServices;
+    try {
+      boolean hasPermissions = LeanplumManifestHelper.checkPermission(LeanplumManifestHelper.RECEIVE_PERMISSION, false, true)
+          && (LeanplumManifestHelper.checkPermission(context.getPackageName() + ".gcm.permission.C2D_MESSAGE", true, false)
+          || LeanplumManifestHelper.checkPermission(context.getPackageName() + ".permission.C2D_MESSAGE", true, true));
+
+      boolean hasGcmReceiver = LeanplumManifestHelper.checkComponent(
+          LeanplumManifestHelper.ApplicationComponent.RECEIVER, LeanplumManifestHelper.GCM_RECEIVER,
+          true, LeanplumManifestHelper.SEND_PERMISSION, Arrays.asList(LeanplumManifestHelper.RECEIVE_ACTION,
+              LeanplumManifestHelper.REGISTRATION_ACTION), context.getPackageName());
+      boolean hasPushReceiver = LeanplumManifestHelper.checkComponent(LeanplumManifestHelper.ApplicationComponent.RECEIVER,
+          LeanplumPushReceiver.class.getName(), false, null,
+          Collections.singletonList(LeanplumManifestHelper.PUSH_LISTENER_SERVICE_FILTER), context.getPackageName());
+
+      boolean hasReceivers = hasGcmReceiver && hasPushReceiver;
+
+      boolean hasPushListenerService = LeanplumManifestHelper.checkComponent(
+          LeanplumManifestHelper.ApplicationComponent.SERVICE,
+          LeanplumPushListenerService.class.getName(), false, null,
+          Collections.singletonList(LeanplumManifestHelper.RECEIVE_ACTION), context.getPackageName());
+      boolean hasInstanceIdService = LeanplumManifestHelper.checkComponent(
+          LeanplumManifestHelper.ApplicationComponent.SERVICE,
+          LeanplumPushInstanceIDService.class.getName(), false, null,
```

Comment:
We believe that `LeanplumPushInstanceIDService.class.getName()`(e.g. trying to reference the actual class here) could lead to a ClassNotFoundException (or similar). We have received exception emails for this before.
The issue is likely that the import of the class fails because it extends from a class that is in the gym package. So when the client doesn't include the gcm package in their app it would fail.

Reply:
This should be testable as well.